### PR TITLE
`ToolsPanel`: refactor Storybook examples to TypeScript, fix types inconsistencies

### PR DESCRIPTION
**Pull request by ciampo**
## What?

Refactor `ToolsPanel` Storybook examples to TypeScript, and iron out some type inconsistencies in the process.

## Why?

Part of #35744 

### Here's a detailed list of type inconsistencies / fixes:

 - Allowed `null` as a possible value for `panelId` on both `ToolsPanel` and `ToolsPanelItem` (as already explained in the `ToolsPanel` README on `trunk`) (https://github.com/WordPress/gutenberg/pull/47944/commits/800a5e5449c72a1815eac73f3c624be137525e58)
 - Marked the `hasInnerWrapper` and `shouldRenderPlaceholderItems` props on `ToolsPanel` as optional (with a default value of `false`). Until now they were marked as required, but they were not passed in many examples (including Storybook) (https://github.com/WordPress/gutenberg/pull/47944/commits/ebe5493a2a0f83e7b68784d1405eed1cfa2727ad)
 - assigned a default value of `noop` to `resetAllFilter`, since TypeScript would otherwise error about it potentially being `undefined` when used in a `useCallback` hook (https://github.com/WordPress/gutenberg/pull/47944/commits/378ee00b4832db1cc8f2dc40095df865a6740827)
 - Marked the `isShownByDefault` prop on `ToolsPanelItem` as optional with a default value of `false`. Until now it was marked as required, but it was not passed in many examples (including Storybook) https://github.com/WordPress/gutenberg/pull/47944/commits/284c5ae405644313f6abc0cb6ade78455ba731af)

### And this is a high-level overview of Storybook changes:

- Renamed files, fixed imports
- Added meta object, including controls settings
- Switched to latest story format using templates and args (also refactored the extra `isFirstToolsPanelItemShownByDefault` prop from args to internal state)
- Added types to all components and hooks that needed them
- Left inline comments where things still need adjusting

### TODO before merging:

- Update JSDoc example snippets if/where necessary
- Check usages of the component around the project, and make sure that they reflect the prop types (value types, requires/optional)

## Testing Instructions

**The only runtime changes are the assignment of some default values for the props that have been marked as optional.**

- Review code changes
- Make sure that the Storybook examples still work as on `trunk`

### CODE REVIEW

File: packages/components/src/tools-panel/stories/index.tsx
Comment (ciampo) [inline packages/components/src/tools-panel/stories/index.tsx:70]: - Making sure that all `props` are passed to `ToolsPanel`. The `label` prop has been moved to the `Story.args` object, instead of being hardcoded in the render function.
- The `resetAll` prop is also passed (see a few lines above in the local `resetAll` function), so that Storybook can correctly fire an action (in the actions tab) when the callback is invoked.

These changes have been applied to all stories.

File: packages/components/src/tools-panel/stories/index.tsx
Comment (ciampo) [inline packages/components/src/tools-panel/stories/index.tsx:223]: This is hacky, but not really in the scope of this PR to change. I just added a comment to add more context around it

File: packages/components/src/tools-panel/stories/index.tsx
Comment (ciampo) [inline packages/components/src/tools-panel/stories/index.tsx:166]: Since `isShownByDefault` is not a prop of `ToolsPanel`, in order to adhere to how controls are supposed to work I decided to move it inside the render function as internal state (`isFirstToolsPanelItemShownByDefault`). Its value can still be changed interactively, thanks to a `button` that I added to the story

File: packages/components/src/tools-panel/stories/index.tsx
Comment (ciampo) [inline packages/components/src/tools-panel/stories/index.tsx:232]: Moved to be part of each story's `args` object

File: packages/components/src/tools-panel/stories/index.tsx
Comment (ciampo) [inline packages/components/src/tools-panel/stories/index.tsx]: This is something that I wasn't able to fix on my own. According to the type defs, `ResetAllFilter` is of type `() => void;`, and therefore is not supposed to accept arguments.

We have two alternatives here:
- either change the `ResetAllFilter` type to accept arguments
- or change the next line to `...resetFilter()`

This same issue happens a few times across the file

File: packages/components/src/tools-panel/stories/utils/tools-panel-with-item-group-slot.tsx
Comment (ciampo) [inline packages/components/src/tools-panel/stories/utils/tools-panel-with-item-group-slot.tsx]: Moved these from being hardcoded values, to being parameters (passed via `ToolsPanelWithItemGroupSlot.args`).

I then had to add prop types for `ToolsPanelItems.Slot`, although this is still quite messy due to the mixing of `ToolsPanel`-specific props with other generic (`any`) props that get forwarded to `Slot`

File: packages/components/src/tools-panel/stories/utils/tools-panel-with-item-group-slot.tsx
Comment (ciampo) [inline packages/components/src/tools-panel/stories/utils/tools-panel-with-item-group-slot.tsx]: I'm not quite sure how to type this. I think that the `as` prop should be generally used with vanilla HTML elements, and we've been abusing it by passing React components.

Maybe we could avoid the `as` prop, and instead do something like

```
<ToolsPanelItems.Slot>
  <ItemGroup>
```

? 

Although this may cause the old `ItemGroup` issue where it needs to be a direct descendant of `Item`

File: packages/components/src/tools-panel/stories/index.tsx
Comment (ciampo) [inline packages/components/src/tools-panel/stories/index.tsx:139]: Ideally we should pass `children` to `ToolsPanel` via the `children` prop in the `args` object. But given how intertwined `ToolsPanel` and `ToolsPanelItem` components are in these stories, for now I've decided to keep them as they were

**Comment by ciampo** [comment]
Hey @aaronrobertshaw , could you take a good look at this PR and let me know if these changes make sense, in the context of `ToolsPanel` ?

I left a detailed list of what's in this PR in the description above. I've also left inline comments — some of them go in details about the changes that I've applied, but others are pointing at parts of the PR that are still incomplete.

If you want to push changes to this branch, feel free to take over the PR and do so!

Thank you 🙏

File: packages/components/src/tools-panel/stories/utils/tools-panel-with-item-group-slot.tsx
Comment (aaronrobertshaw) [reply packages/components/src/tools-panel/stories/utils/tools-panel-with-item-group-slot.tsx]: From memory, this story was around attempts to render color controls as an `ItemGroup` to match what was in Global Styles. 

We couldn't make the `ToolsPanel` the `ItemGroup` or use the `ItemGroup` as an immediate child of the panel because a virtual bubbling slot adds an extra div. Adding an inner `ItemGroup` within the color panel's slot container would also introduce markup that could break the panel layout.

That led to this example to trial rendering the Slot as an `ItemGroup`. I believe we went in a different direction as the fills for a panel could be anything, where the `ItemGroup` would expect them to be an `Item`.

Some further history can be found in https://github.com/WordPress/gutenberg/pull/34027 & https://github.com/WordPress/gutenberg/pull/37546.

Given the issue you've raised and that we didn't end up using the story's approach, would it be better to delete the story entirely?

File: packages/components/src/tools-panel/stories/index.tsx
Comment (aaronrobertshaw) [reply packages/components/src/tools-panel/stories/index.tsx]: Thanks for catching this one 👍 

In terms of which option, I believe the type is incorrect as prior uses of the `resetAllFilter` are to pass through the current data so each filter in turn was aware of the latest changes and wouldn't override updates from previous filters.

If I recall correctly, the color supports panel uses a different batch approach that might allow the second option you presented. 

Given there might already be external use of the ToolsPanel, I think I'd still lean towards fixing the type to match existing uses. Also, the ToolsPanel wasn't typed initially and was converted to TypeScript after the fact, which might further point to the types being incorrect.

File: packages/components/src/tools-panel/stories/index.tsx
Comment (aaronrobertshaw) [inline packages/components/src/tools-panel/stories/index.tsx]:
```suggestion
		resetAllProp( filters );
```

Small typo but it breaks the reset functionality.

File: packages/components/src/tools-panel/stories/index.tsx
Comment (aaronrobertshaw) [inline packages/components/src/tools-panel/stories/index.tsx]:
```suggestion
		resetAllProp( filters );
```

Think this should be `resetAllProp`. The reset functionality is broken otherwise.

**Review by aaronrobertshaw — COMMENTED** [review]
Great work here @ciampo ✨ 

This is looking really good so far. It's great to see the ToolsPanel's kinks being ironed out and the component brought up to scratch.

In general, it tests well for me. The only issues I spotted were a couple of minor copy-and-paste errors in the storybook examples `resetAll` functions.

I've also left a couple of replies to inline comments and questions that might allow us to fix the incorrect `resetAllFilter` type and cull a very edge case story.

✅ Unit tests pass & no typing errors
✅ No issues found while smoke testing in editors
❓ Only minor reset all issue in storybook examples. Once fixed locally, I couldn't spot any other issues with them.

File: packages/components/src/tools-panel/stories/utils/tools-panel-with-item-group-slot.tsx
Comment (ciampo) [reply packages/components/src/tools-panel/stories/utils/tools-panel-with-item-group-slot.tsx]: If the purpose of this story was mainly to showcase the usage with `Item` and `ItemGroup` (and we didn't end up using that approach anyway), then yes — let's delete it :)

File: packages/components/src/tools-panel/stories/index.tsx
Comment (ciampo) [reply packages/components/src/tools-panel/stories/index.tsx]: https://github.com/WordPress/gutenberg/pull/47944/commits/110b82356098a7b22b5533575b2ff2e423789e4f

File: packages/components/src/tools-panel/stories/index.tsx
Comment (ciampo) [reply packages/components/src/tools-panel/stories/index.tsx]: https://github.com/WordPress/gutenberg/pull/47944/commits/110b82356098a7b22b5533575b2ff2e423789e4f

File: packages/components/src/tools-panel/stories/index.tsx
Comment (ciampo) [reply packages/components/src/tools-panel/stories/index.tsx]: > I think I'd still lean towards fixing the type to match existing uses

Ok! I tried to find a good way to describe the `ResetAllFIlter` type , but the fact that it could literally receive any data as an argument A(or undefined), and return any data (or `void`) made me lean towards using `( attributes?: any ) => any` (https://github.com/WordPress/gutenberg/pull/47944/commits/364fb8727711c9bded7678d9dfc833520dfd0859)

(@mirka or @tyxla , do you have any suggestions on how to type this callback more strictly ?)

I've also updated the README in both `ToolsPanel` and `ToolsPanelItem` to reflect those changes (https://github.com/WordPress/gutenberg/pull/47944/commits/08e55bc9cc712d747e3bd1ea6957a890b29c8cf5)

@aaronrobertshaw , what do you think? If this looks good, I'll let you decide what pattern we should advocate for in example snippets (and potentially change usages in the repo to follow that style too).

File: packages/components/src/tools-panel/stories/utils/tools-panel-with-item-group-slot.tsx
Comment (ciampo) [reply packages/components/src/tools-panel/stories/utils/tools-panel-with-item-group-slot.tsx]: Done in https://github.com/WordPress/gutenberg/pull/47944/commits/f7d9bec6eef5d884a516cce8c236b87f1cde39cb

**Comment by ciampo** [comment]
Hey @aaronrobertshaw , thank you for the review and the extra context!

I went ahead and addressed all feedback — all type errors should be gone!

The only "pending" item is the `ResetAllFilter` type (see [convo here](https://github.com/WordPress/gutenberg/pull/47944#discussion_r1102666971)).

Apart from that, it would be great if as part of your final review you could:

- double check that code, types.ts and READMEs are all in sync
- double check that code snippets in JDDocs and READMEs reflect a correct usage of the components, especially after the tweaks included in this PR
- double check that usages in the rest of the repo also follow the current types correctly

Thank you for your patience 🙇

File: packages/components/src/tools-panel/stories/index.tsx
Comment (aaronrobertshaw) [reply packages/components/src/tools-panel/stories/index.tsx]: > what do you think?

I think this is still an improvement over the incorrect type not matching the callback's intended use. 

Unfortunately, I'm short on ideas for typing it more strictly. I'd suggest the general expectation by consumers would be that the same type would be maintained as the data gets passed through each `resetAllFilter`. I don't know how we'd achieve that, though. 

> I'll let you decide what pattern we should advocate for in example snippets (and potentially change usages in the repo to follow that style too).

We have two use cases around resetting values, and only one needs the reset all filters.
1. When the ToolsPanel is created, it has a fixed set of child `ToolsPanelItems`, which aren't rendered via slot fills into the panel. This means the `ToolsPanel` can know what needs to be reset or executed for the `resetAll` function and can avoid individual items needing to define `resetAllFilter`. The best examples here are the Global Styles panels in the site editor e.g. [Dimensions Panel](https://github.com/WordPress/gutenberg/blob/ba1a2ba66f208915edf59f78abd05c4cfc1f60b3/packages/edit-site/src/components/global-styles/dimensions-panel.js#L407-L420).
2. The `ToolsPanel` renders a slot that may contain an unknown number of child `ToolsPanelItems`. In this case, the consumer of the `ToolsPanel` needs to rely on each `ToolsPanelItem` specifying how to effect its reset via the `resetAllFilter` callback. The primary examples of this are the [block support panels](https://github.com/WordPress/gutenberg/blob/8675f86f885ca389f3bca39971b22cb897bab61b/packages/block-editor/src/components/inspector-controls/block-support-tools-panel.js#L24) in the block editor.

Outside of our editors and their block support panels, I believe the first use case above would be the most common. We'd probably be best served to lean towards demonstrating it in our JSDoc example snippets. The Storybook is there to demonstrate the more edge case use of `SlotFills` with the `ToolsPanel`.

I'll push a commit shortly that might form the basis for our example snippet.

Comment (aaronrobertshaw) [reply packages/components/src/tools-panel/stories/index.tsx]: It might also be worth noting that the passing of the `attributes` arg to the `resetAllFilter` only came about due to needing to reset only specific properties within nested objects, e.g. only radius, and not color, width, and style, within a `style.border` attribute. 

Keeping any example use of the `ToolsPanel` simple sounds like a win. If a consumer's use case is more complicated the extra Storybook examples provide direction.

**Review by aaronrobertshaw — APPROVED** [review]
This is testing well for me on all fronts now; nice work 👍 

✅ Unit tests pass & no typing errors
✅ No issues found while smoke testing in editors
✅ Storybook examples function as expected

> Apart from that, it would be great if as part of your final review you could:
> - double check that code, types.ts and READMEs are all in sync

✅ I didn't spot anything amiss between the code, types.ts, and READMEs

> double check that code snippets in JDDocs and READMEs reflect a correct usage of the components, especially after the tweaks included in this PR

✅ JSDoc and README examples appear correct to me. It was more the types that were fixed to match existing usage.

> double check that usages in the rest of the repo also follow the current types correctly

✅ Manually went through `ToolsPanel` uses in Gutenberg and didn't spot any misuse of components or incorrect types.

---

Before giving my final review, I tweaked the export of the `ToolsPanel` and `ToolsPanelItem` components from their respective `component.tsx` files so that their JSDoc descriptions were included in the Storybook docs. (https://github.com/WordPress/gutenberg/pull/47944/commits/c0c86d00ab67ab71a49ac591cd03fc39aeb931ba) 

I also added a JSDoc example for the `ToolsPanel`. That might be worth double checking before merging. (https://github.com/WordPress/gutenberg/pull/47944/commits/b4ec3017034b890d99e6873d9297b46b94161606)

From my end though, this looks good to go 🚢

File: packages/components/src/tools-panel/stories/index.tsx
Comment (ciampo) [reply packages/components/src/tools-panel/stories/index.tsx]: Agreed, let's showcase the simple usage (ie. without `resetAllFilter`) in our example snippets

**Comment by ciampo** [comment]
> Before giving my final review, I tweaked the export of the ToolsPanel and ToolsPanelItem components from their respective component.tsx files so that their JSDoc descriptions were included in the Storybook docs. (https://github.com/WordPress/gutenberg/commit/c0c86d00ab67ab71a49ac591cd03fc39aeb931ba)

I missed that, thank you for pushing that change!

> I also added a JSDoc example for the ToolsPanel. That might be worth double checking before merging. (https://github.com/WordPress/gutenberg/commit/b4ec3017034b890d99e6873d9297b46b94161606)

Looks good too!

> This is testing well for me on all fronts now; nice work 👍
> From my end though, this looks good to go 🚢

Awesome, thank you! I'll plan on opening a PR to refactor unit tests too soon, so that the whole component gets refactored to TypeScript.

**Pull request merged**
